// igniteui-webcomponents/components/progress/circular-gradient.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { registerComponent } from '../common/definitions/register.js';
class IgcCircularGradientComponent extends LitElement {
    constructor() {
        super(...arguments);
        this.offset = '0%';
        this.color = 'black';
        this.opacity = 1;
    }
    static register() {
        registerComponent(IgcCircularGradientComponent);
    }
    render() {
        return nothing;
    }
}
IgcCircularGradientComponent.tagName = 'igc-circular-gradient';
export default IgcCircularGradientComponent;
__decorate([
    property()
], IgcCircularGradientComponent.prototype, "offset", void 0);
__decorate([
    property()
], IgcCircularGradientComponent.prototype, "color", void 0);
__decorate([
    property({ type: Number })
], IgcCircularGradientComponent.prototype, "opacity", void 0);
