// apex-grid/dist/src/controllers/data-operation.js
import SortDataOperation from '../operations/sort.js';
import FilterDataOperation from '../operations/filter.js';
import { isDefined } from '../internal/is-defined.js';
export class DataOperationsController {
    constructor(host) {
        this.host = host;
        this.sorting = new SortDataOperation();
        this.filtering = new FilterDataOperation();
        this.host.addController(this);
    }
    hostConnected() { }
    get hasCustomSort() {
        var _a;
        return isDefined((_a = this.host.dataPipelineConfiguration) === null || _a === void 0 ? void 0 : _a.sort);
    }
    get hasCustomFilter() {
        var _a;
        return isDefined((_a = this.host.dataPipelineConfiguration) === null || _a === void 0 ? void 0 : _a.filter);
    }
    get customFilter() {
        return this.host.dataPipelineConfiguration.filter;
    }
    get customSort() {
        return this.host.dataPipelineConfiguration.sort;
    }
    async apply(data, state) {
        const { filtering, sorting } = state;
        data = this.hasCustomFilter
            ? await this.customFilter({ data, grid: this.host, type: 'filter' })
            : this.filtering.apply(data, filtering.state);
        data = this.hasCustomSort
            ? await this.customSort({ data, grid: this.host, type: 'sort' })
            : this.sorting.apply(data, sorting.state);
        return data;
    }
}
