// igniteui-webcomponents/components/combo/controllers/data.js
import FilterDataOperation from '../operations/filter.js';
import GroupDataOperation from '../operations/group.js';
export class DataController {
    constructor(host) {
        this.host = host;
        this.grouping = new GroupDataOperation();
        this.filtering = new FilterDataOperation();
        this._searchTerm = '';
        this._compareCollator = new Intl.Collator();
        this.host.addController(this);
    }
    set searchTerm(value) {
        this._searchTerm = value;
        this.host.requestUpdate('pipeline');
    }
    get searchTerm() {
        return this._searchTerm;
    }
    get filteringOptions() {
        return this.host.filteringOptions;
    }
    get groupingOptions() {
        return {
            valueKey: this.host.valueKey,
            displayKey: this.host.displayKey,
            groupKey: this.host.groupKey,
            direction: this.host.groupSorting,
        };
    }
    get compareCollator() {
        return this._compareCollator;
    }
    index(data) {
        return data.map((item, index) => ({
            value: item,
            header: false,
            dataIndex: index,
        }));
    }
    hostConnected() { }
    async apply(data) {
        let records = this.index(data);
        records = this.filtering.apply(records, this);
        records = this.grouping.apply(records, this);
        return records;
    }
}
