// apex-grid/dist/src/operations/filter/state.js
import { FilterExpressionTree } from './tree.js';
export class FilterState {
    constructor() {
        this.state = new Map();
    }
    get empty() {
        return this.state.size < 1;
    }
    get keys() {
        return Array.from(this.state.keys());
    }
    get values() {
        return Array.from(this.state.values());
    }
    get ands() {
        return this.values.map(each => each.ands).flat();
    }
    get ors() {
        return this.values.map(each => each.ors).flat();
    }
    has(key) {
        return this.state.has(key);
    }
    get(key) {
        return this.state.get(key);
    }
    delete(key) {
        return this.state.delete(key);
    }
    clear() {
        this.state.clear();
    }
    set(expression) {
        var _a;
        if (this.has(expression.key)) {
            (_a = this.get(expression.key)) === null || _a === void 0 ? void 0 : _a.add(expression);
            return;
        }
        this.state.set(expression.key, new FilterExpressionTree(expression.key).add(expression));
    }
}
