// igniteui-webcomponents/components/badge/badge.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var IgcBadgeComponent_1;
import { LitElement, html } from 'lit';
import { property } from 'lit/decorators.js';
import { themes } from '../../theming/theming-decorator.js';
import { watch } from '../common/decorators/watch.js';
import { registerComponent } from '../common/definitions/register.js';
import { styles } from './themes/badge.base.css.js';
import { styles as shared } from './themes/shared/badge.common.css.js';
import { all } from './themes/themes.js';
let IgcBadgeComponent = IgcBadgeComponent_1 = class IgcBadgeComponent extends LitElement {
    static register() {
        registerComponent(IgcBadgeComponent_1);
    }
    constructor() {
        super();
        this.variant = 'primary';
        this.outlined = false;
        this.shape = 'rounded';
        this.__internals = this.attachInternals();
        this.__internals.role = 'status';
    }
    variantChange() {
        this.__internals.ariaRoleDescription = `badge ${this.variant}`;
    }
    render() {
        return html `
      <span part="base">
        <slot></slot>
      </span>
    `;
    }
};
IgcBadgeComponent.tagName = 'igc-badge';
IgcBadgeComponent.styles = [styles, shared];
__decorate([
    property({ reflect: true })
], IgcBadgeComponent.prototype, "variant", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcBadgeComponent.prototype, "outlined", void 0);
__decorate([
    property({ reflect: true })
], IgcBadgeComponent.prototype, "shape", void 0);
__decorate([
    watch('variant')
], IgcBadgeComponent.prototype, "variantChange", null);
IgcBadgeComponent = IgcBadgeComponent_1 = __decorate([
    themes(all)
], IgcBadgeComponent);
export default IgcBadgeComponent;
