// @lit-labs/context/development/lib/value-notifier.js
/**
 * @license
 * Copyright 2021 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
/**
 * A simple class which stores a value, and triggers registered callbacks when the
 * value is changed via its setter.
 *
 * An implementor might use other observable patterns such as MobX or Redux to get
 * behavior like this. But this is a pretty minimal approach that will likely work
 * for a number of use cases.
 */
export class ValueNotifier {
    constructor(defaultValue) {
        this.disposers = new Map();
        this.updateObservers = () => {
            for (const [callback, disposer] of this.disposers) {
                callback(this._value, disposer);
            }
        };
        if (defaultValue !== undefined) {
            this.value = defaultValue;
        }
    }
    get value() {
        return this._value;
    }
    set value(v) {
        this.setValue(v);
    }
    setValue(v, force = false) {
        const update = force || !Object.is(v, this._value);
        this._value = v;
        if (update) {
            this.updateObservers();
        }
    }
    addCallback(callback, subscribe) {
        if (subscribe) {
            if (!this.disposers.has(callback)) {
                this.disposers.set(callback, () => {
                    this.disposers.delete(callback);
                });
            }
            const disposer = this.disposers.get(callback);
            callback(this.value, disposer);
        }
        else {
            callback(this.value);
        }
    }
    clearCallbacks() {
        this.disposers.clear();
    }
}
