// igniteui-webcomponents/components/divider/divider.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var IgcDividerComponent_1;
import { LitElement, html } from 'lit';
import { property } from 'lit/decorators.js';
import { themes } from '../../theming/theming-decorator.js';
import { registerComponent } from '../common/definitions/register.js';
import { styles } from './themes/divider.base.css.js';
import { styles as shared } from './themes/shared/divider.common.css.js';
import { all } from './themes/themes.js';
let IgcDividerComponent = IgcDividerComponent_1 = class IgcDividerComponent extends LitElement {
    static register() {
        registerComponent(IgcDividerComponent_1);
    }
    set vertical(value) {
        this._vertical = Boolean(value);
        this._internals.ariaOrientation = this._vertical
            ? 'vertical'
            : 'horizontal';
    }
    get vertical() {
        return this._vertical;
    }
    constructor() {
        super();
        this._vertical = false;
        this.middle = false;
        this.type = 'solid';
        this._internals = this.attachInternals();
        this._internals.role = 'separator';
        this._internals.ariaOrientation = this._vertical
            ? 'vertical'
            : 'horizontal';
    }
    render() {
        return html ``;
    }
};
IgcDividerComponent.tagName = 'igc-divider';
IgcDividerComponent.styles = [styles, shared];
__decorate([
    property({ type: Boolean, reflect: true })
], IgcDividerComponent.prototype, "vertical", null);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcDividerComponent.prototype, "middle", void 0);
__decorate([
    property({ reflect: true })
], IgcDividerComponent.prototype, "type", void 0);
IgcDividerComponent = IgcDividerComponent_1 = __decorate([
    themes(all)
], IgcDividerComponent);
export default IgcDividerComponent;
