// igniteui-webcomponents/components/button-group/toggle-button.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var IgcToggleButtonComponent_1;
import { LitElement, html } from 'lit';
import { property, query } from 'lit/decorators.js';
import { themes } from '../../theming/theming-decorator.js';
import { addKeyboardFocusRing } from '../common/controllers/focus-ring.js';
import { registerComponent } from '../common/definitions/register.js';
import { partNameMap } from '../common/util.js';
import { styles } from './themes/button.base.css.js';
import { all } from './themes/button.js';
import { styles as shared } from './themes/shared/button/button.common.css.js';
let IgcToggleButtonComponent = IgcToggleButtonComponent_1 = class IgcToggleButtonComponent extends LitElement {
    constructor() {
        super(...arguments);
        this._kbFocus = addKeyboardFocusRing(this);
        this.selected = false;
        this.disabled = false;
    }
    static register() {
        registerComponent(IgcToggleButtonComponent_1);
    }
    focus(options) {
        this._nativeButton.focus(options);
    }
    blur() {
        this._nativeButton.blur();
    }
    click() {
        this._nativeButton.click();
    }
    render() {
        return html `
      <button
        part=${partNameMap({
            toggle: true,
            focused: this._kbFocus.focused,
        })}
        type="button"
        ?disabled=${this.disabled}
        .ariaLabel=${this.ariaLabel}
        aria-pressed=${this.selected}
        aria-disabled=${this.disabled}
        @click=${this._kbFocus.reset}
        @blur=${this._kbFocus.reset}
      >
        <slot></slot>
      </button>
    `;
    }
};
IgcToggleButtonComponent.styles = [styles, shared];
IgcToggleButtonComponent.tagName = 'igc-toggle-button';
IgcToggleButtonComponent.shadowRootOptions = {
    ...LitElement.shadowRootOptions,
    delegatesFocus: true,
};
__decorate([
    query('[part="toggle"]', true)
], IgcToggleButtonComponent.prototype, "_nativeButton", void 0);
__decorate([
    property()
], IgcToggleButtonComponent.prototype, "value", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcToggleButtonComponent.prototype, "selected", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcToggleButtonComponent.prototype, "disabled", void 0);
IgcToggleButtonComponent = IgcToggleButtonComponent_1 = __decorate([
    themes(all)
], IgcToggleButtonComponent);
export default IgcToggleButtonComponent;
